[Inductor][CPP] Fix Index name error

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134645

**Summary**

Fix the comment: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2313930242. For all of the cases we see in the 3 test suits (TorchBench, Timms, Huggingface) we expect:

* `_node` is a FX Node with target in ["index_expr", "load", "store"]
* `_node.args[1 if _node.target == "index_expr" else 2]` is another FX node with target `get_index`
* `_node.args[1 if _node.target == "index_expr" else 2].args[0]` is a str for the name of this index expression

It seems not true in some FB internal testcase from the failure log posted in above link. So, add the condition check to work around it.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

leslie-fang-intel committed:
[Inductor][CPP] Fix Index name error

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @masnesral, could you kindly help to check if it can fix the internal failure? Also, it will be helpful to post a reproduce for this failure to confirm our guess from the failure log.

leslie-fang-intel committed:
Update on "[Inductor][CPP] Fix Index name error"


**Summary**

Fix the comment: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2313930242. For all of the cases we see in the 3 test suits (TorchBench, Timms, Huggingface) we expect:

* `_node` is a FX Node with target in ["index_expr", "load", "store"]
* `_node.args[1 if _node.target == "index_expr" else 2]` is another FX node with target `get_index`
* `_node.args[1 if _node.target == "index_expr" else 2].args[0]` is a str for the name of this index expression

It seems not true in some FB internal testcase from the failure log posted in above link. So, add the condition check to work around it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral commented:
@leslie-fang-intel unfortunately I get the same exception on the internal workload. It just occurs on line 3401 when accessing the `args` attribute. As I mentioned on the other PR, it makes NO SENSE! If I catch the exception and print out the values, I'm seeing that `_node.args[2]` indeed has type `<class 'torch.fx.node.Node'>` -- it's not a string at all.

Furthermore, a no-op change causes the mysterious exception to vanish. The current code raises `Exception: 'str' object has no attribute 'args'`:

```
                                index = sub_block.body.indexing_from_args(
                                    (vars, reduction_vars)
                                )[
                                    _node.args[
                                        1 if _node.target == "index_expr" else 2
                                    ].args[0]
                                ]
```

This code does not raise:
```
                                iii = 1 if _node.target == "index_expr" else 2
                                index = sub_block.body.indexing_from_args(
                                    (vars, reduction_vars)
                                )[
                                    _node.args[
                                        iii
                                    ].args[0]
                                ]
```

leslie-fang-intel committed:
Update on "[Inductor][CPP] Fix Index name error"


**Summary**

Fix the comment: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2313930242. For all of the cases we see in the 3 test suits (TorchBench, Timms, Huggingface) we expect:

* `_node` is a FX Node with target in ["index_expr", "load", "store"]
* `_node.args[1 if _node.target == "index_expr" else 2]` is another FX node with target `get_index`
* `_node.args[1 if _node.target == "index_expr" else 2].args[0]` is a str for the name of this index expression

It seems not true in some FB internal testcase from the failure log posted in above link. So, add the condition check to work around it.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @masnesral, thanks for the investigation. It seems the issue is related to the internal environment. I’ve updated this PR based on your findings. Could you please try again?

masnesral commented:
Well I see no problem landing this since it _should_ be functionally equivalent. I'm a bit worried that I have no idea what's going on here, but I suppose that's sort of independent from this change :(

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
@pytorchbot merge
